src/ManagedHandle.zig

```
@@ -0,0 +1,136 @@
+};
+
+pub fn init(arch: enums.Arch, mode: enums.Mode, options: Options) !Self {
+    const handle = try impl.open(arch, mode);
+
+    if (options.syntax != .DEFAULT) {
+        try impl.option(handle, .SYNTAX, @intFromEnum(options.syntax));
+    }
+
+    if (options.detail) {
+        try impl.option(handle, .DETAIL, cs.CS_OPT_ON);
+    }
+
+    if (options.skipdata) {
+        try impl.option(handle, .SKIPDATA, cs.CS_OPT_ON);
+    }
+
+    if (options.unsigned) {
+        try impl.option(handle, .UNSIGNED, cs.CS_OPT_ON);
+    }
+
+    if (options.no_branch_offset) {
+        try impl.option(handle, .NO_BRANCH_OFFSET, cs.CS_OPT_ON);
+    }
+
+    if (options.mnemonic.len > 0) {
+        try impl.option(handle, .MNEMONIC, @intFromPtr(options.mnemonic.ptr));
+    }
+
+    return Self{ .native = handle };
+}
+
+/// Allocates using the cs_malloc function or the user defined one which is set through cs_option/setup.initCapstone(Manually).
+/// Should be freed using cs_free, capstone provided free function or the user defined one if set through option.
+pub fn disasm(self: Self, code: []const u8, address: usize, count: usize) err.CapstoneError![]insn.Insn {
+    return impl.disasm(self.native, code, address, count);
+}
+
+/// Copies using zig user-land allocator and frees the cs_malloc'd memory.
+pub fn disasmAlloc(self: Self, allocator: std.mem.Allocator, code: []const u8, address: usize, count: usize) ![]insn.Insn {
+    const disass = try impl.disasm(self.native, code, address, count);
+    defer impl.free(disass);
+    return allocator.dupe(insn.Insn, disass);
+}
+
+/// Uses global/threadlocal storage to avoid heap allocations for the single Insn struct.
+pub fn disasmIter(self: Self, code: []const u8, address: u64) Iter {
+    return Iter.init(self.native, code, address, @ptrCast(TmpStorage.getIns()));
+}
+
+pub fn deinit(self: *Self) void {
+    impl.close(&self.native) catch |e| {
+        std.debug.print("Failed to close handle: {any}\n", .{impl.strerror(e)});
```

don't use debug print here, instead:
```zig
std.io.getStdErr().writer().print("Failed to close handle: {any}\n", .{impl.strerror(e)});
```

---

hmm so `_ = std.io.getStdErr().writer().print("Failed to close handle: {any}\n", .{impl.strerror(e)});` since the `print` returns error, I don't think a `deinit` function should return any error?

---

correct. :D